Already disposed: Project (Disposed) sqldelight

```
Already disposed: Project (Disposed) sqldelight
java.lang.AssertionError: Already disposed: Project (Disposed) sqldelight
	at com.intellij.openapi.components.impl.ComponentManagerImpl.getComponent(ComponentManagerImpl.java:170)
	at com.intellij.psi.PsiDocumentManager.getInstance(PsiDocumentManager.java:48)
	at com.intellij.psi.SingleRootFileViewProvider.getLastCommittedText(SingleRootFileViewProvider.java:637)
	at com.intellij.psi.SingleRootFileViewProvider.access$300(SingleRootFileViewProvider.java:67)
	at com.intellij.psi.SingleRootFileViewProvider$VirtualFileContent.getText(SingleRootFileViewProvider.java:605)
	at com.intellij.psi.SingleRootFileViewProvider.getContents(SingleRootFileViewProvider.java:424)
	at com.intellij.psi.impl.source.PsiFileImpl.getText(PsiFileImpl.java:386)
	at com.squareup.sqldelight.intellij.lang.SqliteFile.parseThen(SqliteFile.kt:45)
	at com.squareup.sqldelight.intellij.lang.SqlDelightFileViewProvider.generateJavaInterface$sqldelight_studio_plugin_compileKotlin(SqlDelightFileViewProviderFactory.kt:60)
	at com.squareup.sqldelight.intellij.lang.SqlDelightFileViewProvider.generateJavaInterface$sqldelight_studio_plugin_compileKotlin$default(SqlDelightFileViewProviderFactory.kt:59)
	at com.squareup.sqldelight.intellij.lang.SqlDelightFileViewProvider$Companion.triggerDependencyRefresh(SqlDelightFileViewProviderFactory.kt:124)
	at com.squareup.sqldelight.intellij.lang.SqlDelightFileViewProvider$Companion.removeFile(SqlDelightFileViewProviderFactory.kt:117)
	at com.squareup.sqldelight.intellij.lang.SqlDelightFileViewProvider$generateJavaInterface$2.invoke(SqlDelightFileViewProviderFactory.kt:100)
	at com.squareup.sqldelight.intellij.lang.SqlDelightFileViewProvider$generateJavaInterface$2.invoke(SqlDelightFileViewProviderFactory.kt:45)
	at com.squareup.sqldelight.intellij.lang.SqliteFile.parseThen(SqliteFile.kt:57)
	at com.squareup.sqldelight.intellij.lang.SqlDelightFileViewProvider.generateJavaInterface$sqldelight_studio_plugin_compileKotlin(SqlDelightFileViewProviderFactory.kt:60)
	at com.squareup.sqldelight.intellij.lang.SqlDelightFileViewProvider$contentsSynchronized$1.run(SqlDelightFileViewProviderFactory.kt:56)
	at com.intellij.psi.impl.PsiDocumentManagerBase.performWhenAllCommitted(PsiDocumentManagerBase.java:489)
	at com.squareup.sqldelight.intellij.lang.SqlDelightFileViewProvider.contentsSynchronized(SqlDelightFileViewProviderFactory.kt:56)
	at com.intellij.psi.impl.PsiDocumentManagerBase.finishCommitInWriteAction(PsiDocumentManagerBase.java:346)
	at com.intellij.psi.impl.PsiDocumentManagerImpl.finishCommitInWriteAction(PsiDocumentManagerImpl.java:143)
	at com.intellij.psi.impl.PsiDocumentManagerBase$3.run(PsiDocumentManagerBase.java:298)
	at com.intellij.openapi.application.impl.ApplicationImpl.runWriteAction(ApplicationImpl.java:1010)
	at com.intellij.psi.impl.PsiDocumentManagerBase.finishCommit(PsiDocumentManagerBase.java:295)
	at com.intellij.psi.impl.DocumentCommitThread$5.run(DocumentCommitThread.java:528)
	at com.intellij.openapi.application.impl.LaterInvocator$FlushQueue.runNextEvent(LaterInvocator.java:337)
	at com.intellij.openapi.application.impl.LaterInvocator$FlushQueue.run(LaterInvocator.java:321)
	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:209)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:715)
	at java.awt.EventQueue.access$400(EventQueue.java:82)
	at java.awt.EventQueue$2.run(EventQueue.java:676)
	at java.awt.EventQueue$2.run(EventQueue.java:674)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.AccessControlContext$1.doIntersectionPrivilege(AccessControlContext.java:86)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:685)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:866)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:654)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:381)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:296)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:211)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:201)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:196)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:188)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:122)



java.lang.AssertionError: Already disposed: Project (Disposed) sqldelight
	at com.intellij.openapi.components.impl.ComponentManagerImpl.getComponent(ComponentManagerImpl.java:170)
	at com.intellij.psi.PsiDocumentManager.getInstance(PsiDocumentManager.java:48)
	at com.intellij.psi.SingleRootFileViewProvider.getLastCommittedText(SingleRootFileViewProvider.java:637)
	at com.intellij.psi.SingleRootFileViewProvider.access$300(SingleRootFileViewProvider.java:67)
	at com.intellij.psi.SingleRootFileViewProvider$VirtualFileContent.getText(SingleRootFileViewProvider.java:605)
	at com.intellij.psi.SingleRootFileViewProvider.getContents(SingleRootFileViewProvider.java:424)
	at com.intellij.psi.impl.source.PsiFileImpl.getText(PsiFileImpl.java:386)
	at com.squareup.sqldelight.intellij.lang.SqliteFile.parseThen(SqliteFile.kt:45)
	at com.squareup.sqldelight.intellij.lang.SqlDelightFileViewProvider.generateJavaInterface$sqldelight_studio_plugin_compileKotlin(SqlDelightFileViewProviderFactory.kt:60)
	at com.squareup.sqldelight.intellij.lang.SqlDelightFileViewProvider.generateJavaInterface$sqldelight_studio_plugin_compileKotlin$default(SqlDelightFileViewProviderFactory.kt:59)
	at com.squareup.sqldelight.intellij.lang.SqlDelightFileViewProvider$Companion.triggerDependencyRefresh(SqlDelightFileViewProviderFactory.kt:124)
	at com.squareup.sqldelight.intellij.lang.SqlDelightFileViewProvider$Companion.access$triggerDependencyRefresh(SqlDelightFileViewProviderFactory.kt:104)
	at com.squareup.sqldelight.intellij.lang.SqlDelightFileViewProvider$generateJavaInterface$1.invoke(SqlDelightFileViewProviderFactory.kt:74)
	at com.squareup.sqldelight.intellij.lang.SqlDelightFileViewProvider$generateJavaInterface$1.invoke(SqlDelightFileViewProviderFactory.kt:45)
	at com.squareup.sqldelight.intellij.lang.SqliteFile.parseThen(SqliteFile.kt:62)
	at com.squareup.sqldelight.intellij.lang.SqlDelightFileViewProvider.generateJavaInterface$sqldelight_studio_plugin_compileKotlin(SqlDelightFileViewProviderFactory.kt:60)
	at com.squareup.sqldelight.intellij.lang.SqlDelightFileViewProvider$contentsSynchronized$1.run(SqlDelightFileViewProviderFactory.kt:56)
	at com.intellij.psi.impl.PsiDocumentManagerBase.performWhenAllCommitted(PsiDocumentManagerBase.java:489)
	at com.squareup.sqldelight.intellij.lang.SqlDelightFileViewProvider.contentsSynchronized(SqlDelightFileViewProviderFactory.kt:56)
	at com.intellij.psi.impl.PsiDocumentManagerBase.finishCommitInWriteAction(PsiDocumentManagerBase.java:346)
	at com.intellij.psi.impl.PsiDocumentManagerImpl.finishCommitInWriteAction(PsiDocumentManagerImpl.java:143)
	at com.intellij.psi.impl.PsiDocumentManagerBase$3.run(PsiDocumentManagerBase.java:298)
	at com.intellij.openapi.application.impl.ApplicationImpl.runWriteAction(ApplicationImpl.java:1010)
	at com.intellij.psi.impl.PsiDocumentManagerBase.finishCommit(PsiDocumentManagerBase.java:295)
	at com.intellij.psi.impl.DocumentCommitThread$5.run(DocumentCommitThread.java:528)
	at com.intellij.openapi.application.impl.LaterInvocator$FlushQueue.runNextEvent(LaterInvocator.java:337)
	at com.intellij.openapi.application.impl.LaterInvocator$FlushQueue.run(LaterInvocator.java:321)
	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:209)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:715)
	at java.awt.EventQueue.access$400(EventQueue.java:82)
	at java.awt.EventQueue$2.run(EventQueue.java:676)
	at java.awt.EventQueue$2.run(EventQueue.java:674)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.AccessControlContext$1.doIntersectionPrivilege(AccessControlContext.java:86)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:685)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:866)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:654)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:381)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:296)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:211)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:201)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:196)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:188)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:122)
```

happened after checking out a different branch from git, IDE has also been open real long. Restarting IDE fixes issue.

--- a/sqldelight-studio-plugin/src/main/kotlin/com/squareup/sqldelight/intellij/SqlDelightManager.kt
+++ b/sqldelight-studio-plugin/src/main/kotlin/com/squareup/sqldelight/intellij/SqlDelightManager.kt
@@ -1,0 +1,114 @@
+/*
+ * Copyright (C) 2016 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.sqldelight.intellij
+
+import com.intellij.openapi.module.Module
+import com.intellij.openapi.module.ModuleUtil
+import com.intellij.openapi.vfs.VirtualFile
+import com.intellij.psi.PsiElement
+import com.intellij.util.containers.BidirectionalMap
+import com.intellij.util.containers.MultiMap
+import com.squareup.sqldelight.SqliteParser
+import com.squareup.sqldelight.intellij.lang.SqlDelightFileViewProvider
+import com.squareup.sqldelight.intellij.lang.SqliteFile
+import com.squareup.sqldelight.intellij.util.containingParse
+import com.squareup.sqldelight.types.SymbolTable
+import com.squareup.sqldelight.validation.SqlDelightValidator.Companion.ALL_FILE_DEPENDENCY
+import org.antlr.v4.runtime.ParserRuleContext
+
+internal class SqlDelightManager private constructor() {
+  internal var symbolTable = SymbolTable()
+  private val parseTreeMap = BidirectionalMap<SqliteFile, SqliteParser.ParseContext>()
+  private val dependencies = MultiMap<Any, SqlDelightFileViewProvider>()
+
+  fun setParseTree(file: SqliteFile, parseContext: SqliteParser.ParseContext) {
+    parseTreeMap.remove(file)
+    parseTreeMap.put(file, parseContext)
+  }
+
+  fun getPsi(parsed: ParserRuleContext) = parseTreeMap.getKeysByValue(parsed.containingParse())!!
+      .first().findElementAt(parsed.start.startIndex)
+
+  fun setDependencies(fileViewProvider: SqlDelightFileViewProvider, newDependencies: () -> List<Any>) {
+    synchronized(dependencies) {
+      dependencies.entrySet().forEach {
+        it.value.remove(fileViewProvider)
+      }
+      newDependencies().forEach {
+        dependencies.putValue(it, fileViewProvider)
+      }
+    }
+  }
+
+  fun removeFile(
+      file: VirtualFile,
+      fromEdit: Boolean = false,
+      replacementTable: SymbolTable? = null
+  ) {
+    symbolTable -= file
+    if (replacementTable != null) {
+      symbolTable += replacementTable
+    }
+    dependencies.entrySet().forEach {
+      it.value.removeAll { it.virtualFile == file }
+    }
+    triggerDependencyRefresh(file, fromEdit)
+    dependencies.remove(file)
+  }
+
+  fun triggerDependencyRefresh(file: VirtualFile, fromEdit: Boolean = false) {
+    if (fromEdit) {
+      (dependencies.get(file) + dependencies.get(ALL_FILE_DEPENDENCY))
+          .forEach { it.generateJavaInterface() }
+    }
+  }
+
+  companion object {
+    private val managers = linkedMapOf<Module, SqlDelightManager>()
+
+    fun initManager(module: Module) {
+      synchronized(managers) {
+        managers.put(module, SqlDelightManager())
+      }
+    }
+
+    fun disposeManager(module: Module) {
+      synchronized(managers) {
+        managers.remove(module)
+      }
+    }
+
+    fun getInstance(module: Module): SqlDelightManager? {
+      synchronized(managers) {
+        return managers[module]
+      }
+    }
+
+    fun getInstance(element: PsiElement): SqlDelightManager? {
+      try {
+        return getInstance(ModuleUtil.findModuleForPsiElement(element) ?: return null)
+      } catch (e: Throwable) {
+        return null
+      }
+    }
+
+    fun removeFile(file: VirtualFile) {
+      synchronized(managers) {
+        managers.values.forEach { it.removeFile(file, true) }
+      }
+    }
+  }
+}

--- a/sqldelight-studio-plugin/src/main/kotlin/com/squareup/sqldelight/intellij/SqlDelightModuleComponent.kt
+++ b/sqldelight-studio-plugin/src/main/kotlin/com/squareup/sqldelight/intellij/SqlDelightModuleComponent.kt
@@ -32,20 +32,22 @@ import org.jetbrains.plugins.groovy.lang.psi.GroovyPsiElement
 import javax.swing.event.HyperlinkEvent
 
 class SqlDelightModuleComponent(val module: Module) : ModuleComponent {
-  val properties = PropertiesComponent.getInstance(module.project)
-
   override fun projectClosed() { }
   override fun projectOpened() { }
   override fun moduleAdded() { }
-  override fun disposeComponent() { }
+  override fun disposeComponent() {
+    SqlDelightManager.disposeManager(module)
+  }
   override fun getComponentName() = "SqlDelightModuleComponent"
   override fun initComponent() {
+    SqlDelightManager.initManager(module)
     iterateClasspath(module) { classpath ->
       val classpathValue = BuildFileKeyType.STRING.getValue(classpath) as? String
       if (classpathValue == null || classpathValue == GradleBuildFile.UNRECOGNIZED_VALUE) {
         return@iterateClasspath
       }
 
+      val properties = PropertiesComponent.getInstance(module.project)
       val gradleVersion = classpathValue.substringAfterLast(':')
 
       if (GRADLE_PREFIX != classpathValue.substringBeforeLast(':')

--- a/sqldelight-studio-plugin/src/main/kotlin/com/squareup/sqldelight/intellij/SqlDelightStartupActivity.kt
+++ b/sqldelight-studio-plugin/src/main/kotlin/com/squareup/sqldelight/intellij/SqlDelightStartupActivity.kt
@@ -37,8 +37,9 @@ class SqlDelightStartupActivity : StartupActivity {
           true
         })
     files.forEach { file ->
+      val manager = SqlDelightManager.getInstance(file) ?: return@forEach
       file.parseThen({ parsed ->
-        SqlDelightFileViewProvider.symbolTable += SymbolTable(parsed, file.virtualFile)
+        manager.symbolTable += SymbolTable(parsed, file.virtualFile)
       })
     }
     files.forEach { file ->

--- a/sqldelight-studio-plugin/src/main/kotlin/com/squareup/sqldelight/intellij/SqlDelightVirtualFileListener.kt
+++ b/sqldelight-studio-plugin/src/main/kotlin/com/squareup/sqldelight/intellij/SqlDelightVirtualFileListener.kt
@@ -17,10 +17,9 @@ package com.squareup.sqldelight.intellij
 
 import com.intellij.openapi.vfs.VirtualFileAdapter
 import com.intellij.openapi.vfs.VirtualFileEvent
-import com.squareup.sqldelight.intellij.lang.SqlDelightFileViewProvider
 
 class SqlDelightVirtualFileListener : VirtualFileAdapter() {
   override fun beforeFileDeletion(event: VirtualFileEvent) {
-    SqlDelightFileViewProvider.removeFile(event.file, true)
+    SqlDelightManager.removeFile(event.file)
   }
 }

--- a/sqldelight-studio-plugin/src/main/kotlin/com/squareup/sqldelight/intellij/lang/SqlDelightCompletionContributor.kt
+++ b/sqldelight-studio-plugin/src/main/kotlin/com/squareup/sqldelight/intellij/lang/SqlDelightCompletionContributor.kt
@@ -28,6 +28,7 @@ import com.intellij.patterns.InitialPatternCondition
 import com.intellij.psi.PsiElement
 import com.intellij.util.ProcessingContext
 import com.squareup.sqldelight.SqliteParser
+import com.squareup.sqldelight.intellij.SqlDelightManager
 import com.squareup.sqldelight.types.ResolutionError
 import com.squareup.sqldelight.types.Resolver
 import com.squareup.sqldelight.validation.SqlDelightValidator
@@ -51,11 +52,12 @@ class SqlDelightCompletionContributor : CompletionContributor() {
 private class SqlDelightCompletionProvider : CompletionProvider<CompletionParameters>() {
   override fun addCompletions(parameters: CompletionParameters, context: ProcessingContext?,
       result: CompletionResultSet) {
+    val manager = SqlDelightManager.getInstance(parameters.position) ?: return
     val file = parameters.position.containingFile as SqliteFile
     file.dirty = true
     file.parseThen(
-        operation = getAvailableValues(parameters, result),
-        onError = { parsed, errors -> getAvailableValues(parameters, result).invoke(parsed) }
+        operation = getAvailableValues(parameters, result, manager),
+        onError = { parsed, errors -> getAvailableValues(parameters, result, manager).invoke(parsed) }
     )
     // No reason to do any other completion for SQLDelight files. Might save some time.
     result.stopHere()
@@ -63,13 +65,14 @@ private class SqlDelightCompletionProvider : CompletionProvider<CompletionParame
 
   private fun getAvailableValues(
       parameters: CompletionParameters,
-      result: CompletionResultSet
+      result: CompletionResultSet,
+      manager: SqlDelightManager
   ) = { parsed: SqliteParser.ParseContext ->
     result.addAllElements(parsed.sql_stmt_list().sql_stmt()
         .filter { it.start.startIndex < parameters.offset && it.stop.stopIndex > parameters.offset }
         .flatMap {
           try {
-            SqlDelightValidator().validate(it, Resolver(SqlDelightFileViewProvider.symbolTable))
+            SqlDelightValidator().validate(it, Resolver(manager.symbolTable))
           } catch (e: Throwable) {
             emptyList<ResolutionError>()
           }

--- a/sqldelight-studio-plugin/src/main/kotlin/com/squareup/sqldelight/intellij/lang/SqlDelightFileViewProviderFactory.kt
+++ b/sqldelight-studio-plugin/src/main/kotlin/com/squareup/sqldelight/intellij/lang/SqlDelightFileViewProviderFactory.kt
@@ -24,15 +24,14 @@ import com.intellij.psi.FileViewProviderFactory
 import com.intellij.psi.PsiDocumentManager
 import com.intellij.psi.PsiManager
 import com.intellij.psi.SingleRootFileViewProvider
-import com.intellij.util.containers.MultiMap
 import com.squareup.sqldelight.SqliteCompiler
 import com.squareup.sqldelight.Status
 import com.squareup.sqldelight.Status.ValidationStatus.Invalid
+import com.squareup.sqldelight.intellij.SqlDelightManager
 import com.squareup.sqldelight.model.moduleDirectory
 import com.squareup.sqldelight.model.relativePath
 import com.squareup.sqldelight.types.SymbolTable
 import com.squareup.sqldelight.validation.SqlDelightValidator
-import com.squareup.sqldelight.validation.SqlDelightValidator.Companion.ALL_FILE_DEPENDENCY
 import java.io.File
 
 class SqlDelightFileViewProviderFactory : FileViewProviderFactory {
@@ -57,23 +56,17 @@ internal class SqlDelightFileViewProvider(virtualFile: VirtualFile, language: La
   }
 
   internal fun generateJavaInterface(fromEdit: Boolean = false) {
+    val manager = SqlDelightManager.getInstance(file) ?: return
+
     // Mark the file as dirty and re-parse.
     file.dirty = true
     file.parseThen({ parsed ->
-      symbolTable += SymbolTable(parsed, virtualFile)
-      synchronized(dependencies) {
-        dependencies.entrySet().forEach {
-          it.value.remove(this)
-        }
-        file.status = sqldelightValidator.validate(parsed, symbolTable)
-        (file.status as Status.ValidationStatus).dependencies.forEach {
-          if (it != virtualFile) {
-            dependencies.putValue(it, this)
-          }
-        }
+      manager.symbolTable += SymbolTable(parsed, virtualFile)
+      manager.setDependencies(this) {
+        file.status = sqldelightValidator.validate(parsed, manager.symbolTable)
+        (file.status as Status.ValidationStatus).dependencies.filter { it != virtualFile }
       }
-
-      triggerDependencyRefresh(virtualFile, fromEdit)
+      manager.triggerDependencyRefresh(virtualFile, fromEdit)
 
       if (file.status is Invalid) return@parseThen
 
@@ -99,7 +92,7 @@ internal class SqlDelightFileViewProvider(virtualFile: VirtualFile, language: La
         file.generatedFile = psiManager.findFile(generatedFile)
       }
     }, onError = { parsed, errors ->
-      removeFile(virtualFile, fromEdit, SymbolTable(parsed, virtualFile, errors))
+      manager.removeFile(virtualFile, fromEdit, SymbolTable(parsed, virtualFile, errors))
     })
   }
 
@@ -107,32 +100,6 @@ internal class SqlDelightFileViewProvider(virtualFile: VirtualFile, language: La
     private val localFileSystem = LocalFileSystem.getInstance()
     private val sqliteCompiler = SqliteCompiler()
     private val sqldelightValidator = SqlDelightValidator()
-    private val dependencies = MultiMap<Any, SqlDelightFileViewProvider>()
-
-    internal var symbolTable = SymbolTable()
-
-    fun removeFile(
-        file: VirtualFile,
-        fromEdit: Boolean = false,
-        replacementTable: SymbolTable? = null
-    ) {
-      symbolTable -= file
-      if (replacementTable != null) {
-        symbolTable += replacementTable
-      }
-      dependencies.entrySet().forEach {
-        it.value.removeAll { it.virtualFile == file }
-      }
-      triggerDependencyRefresh(file, fromEdit)
-      dependencies.remove(file)
-    }
-
-    private fun triggerDependencyRefresh(file: VirtualFile, fromEdit: Boolean = false) {
-      if (fromEdit) {
-        (dependencies.get(file) + dependencies.get(ALL_FILE_DEPENDENCY))
-            .forEach { it.generateJavaInterface() }
-      }
-    }
   }
 }
 

--- a/sqldelight-studio-plugin/src/main/kotlin/com/squareup/sqldelight/intellij/lang/SqliteFile.kt
+++ b/sqldelight-studio-plugin/src/main/kotlin/com/squareup/sqldelight/intellij/lang/SqliteFile.kt
@@ -18,11 +18,11 @@ package com.squareup.sqldelight.intellij.lang
 import com.intellij.extapi.psi.PsiFileBase
 import com.intellij.psi.FileViewProvider
 import com.intellij.psi.PsiFile
-import com.intellij.util.containers.BidirectionalMap
 import com.squareup.sqldelight.SqliteLexer
 import com.squareup.sqldelight.SqliteParser
 import com.squareup.sqldelight.SqlitePluginException
 import com.squareup.sqldelight.Status
+import com.squareup.sqldelight.intellij.SqlDelightManager
 import org.antlr.v4.runtime.ANTLRInputStream
 import org.antlr.v4.runtime.BaseErrorListener
 import org.antlr.v4.runtime.CommonTokenStream
@@ -45,7 +45,9 @@ class SqliteFile internal constructor(viewProvider: FileViewProvider)
       operation: (SqliteParser.ParseContext) -> Unit,
       onError: (SqliteParser.ParseContext, List<Token>) -> Unit = { parsed, errors -> /* no op */ }
   ) {
-    synchronized (project) {
+    val manager = SqlDelightManager.getInstance(this) ?: return
+
+    synchronized (manager) {
       if (!dirty) {
         operation(parsed)
         return@synchronized
@@ -62,8 +64,7 @@ class SqliteFile internal constructor(viewProvider: FileViewProvider)
 
       val parsed = parser.parse()
       this.parsed = parsed;
-      parseTreeMap.remove(this)
-      parseTreeMap.put(this, parsed)
+      manager.setParseTree(this, parsed)
 
       if (errorListener.errors.isNotEmpty()) {
         // Syntax level errors are handled by the annotator. Don't generate anything.
@@ -86,9 +87,5 @@ class SqliteFile internal constructor(viewProvider: FileViewProvider)
         charPositionInLine: Int, msg: String?, e: RecognitionException?) {
       errors.add(offendingSymbol as Token)
     }
-  }
-
-  companion object {
-    internal val parseTreeMap = BidirectionalMap<SqliteFile, SqliteParser.ParseContext>()
   }
 }

--- a/sqldelight-studio-plugin/src/main/kotlin/com/squareup/sqldelight/intellij/psi/SqlDelightElementRef.kt
+++ b/sqldelight-studio-plugin/src/main/kotlin/com/squareup/sqldelight/intellij/psi/SqlDelightElementRef.kt
@@ -18,9 +18,8 @@ package com.squareup.sqldelight.intellij.psi
 import com.intellij.openapi.util.TextRange
 import com.intellij.psi.PsiElement
 import com.intellij.psi.PsiReferenceBase
-import com.squareup.sqldelight.intellij.lang.SqlDelightFileViewProvider
+import com.squareup.sqldelight.intellij.SqlDelightManager
 import com.squareup.sqldelight.intellij.lang.SqliteFile
-import com.squareup.sqldelight.intellij.util.containingParse
 import com.squareup.sqldelight.intellij.util.doRename
 import com.squareup.sqldelight.intellij.util.findUsages
 import com.squareup.sqldelight.intellij.util.isDefinition
@@ -37,6 +36,7 @@ class SqlDelightElementRef(idNode: IdentifierElement, private val ruleName: Stri
   override fun getVariants() = emptyArray<Any>()
 
   override fun resolve(): PsiElement? {
+    val manager = SqlDelightManager.getInstance(element) ?: return null
     var result: PsiElement? = null
     (element.containingFile as SqliteFile).parseThen({ parsed ->
       val ruleAtElement = parsed.leafAt(element.textOffset)
@@ -48,7 +48,7 @@ class SqlDelightElementRef(idNode: IdentifierElement, private val ruleName: Stri
           .filter { it.start.startIndex < element.textOffset && it.stop.stopIndex > element.textOffset }
           .flatMap {
             try {
-              SqlDelightValidator().validate(it, Resolver(SqlDelightFileViewProvider.symbolTable,
+              SqlDelightValidator().validate(it, Resolver(manager.symbolTable,
                   elementToFind = element.textOffset))
             } catch (e: Throwable) {
               emptyList<ResolutionError>()
@@ -57,9 +57,7 @@ class SqlDelightElementRef(idNode: IdentifierElement, private val ruleName: Stri
           .filterIsInstance<ResolutionError.ElementFound>()
           .firstOrNull()
       if (elementFound != null) {
-        result = SqliteFile.parseTreeMap.getKeysByValue(
-            elementFound.originatingElement.containingParse())!!.first().findElementAt(
-            elementFound.originatingElement.start.startIndex)
+        result = manager.getPsi(elementFound.originatingElement)
       }
     })
     return result

--- a/sqldelight-studio-plugin/src/test/kotlin/com/squareup/sqldelight/SqlDelightFixtureTestCase.kt
+++ b/sqldelight-studio-plugin/src/test/kotlin/com/squareup/sqldelight/SqlDelightFixtureTestCase.kt
@@ -18,6 +18,7 @@ package com.squareup.sqldelight
 import com.google.common.truth.Truth.assertThat
 import com.intellij.codeInsight.completion.CompletionType
 import com.intellij.testFramework.fixtures.LightPlatformCodeInsightFixtureTestCase
+import com.squareup.sqldelight.intellij.SqlDelightManager
 import com.squareup.sqldelight.intellij.lang.SqlDelightFileViewProvider
 import com.squareup.sqldelight.types.SymbolTable
 import java.io.File
@@ -41,7 +42,7 @@ abstract class SqlDelightFixtureTestCase : LightPlatformCodeInsightFixtureTestCa
     }
 
     // Reset the symbol table.
-    SqlDelightFileViewProvider.symbolTable = SymbolTable()
+    SqlDelightManager.getInstance(myModule)!!.symbolTable = SymbolTable()
 
     // Load universal files into the IDE.
     myFixture.configureByFiles(*filesToLoad.toTypedArray())

